retrieval/transform.go

```diff
@@ -121,21 +123,26 @@ func (b *sampleBuilder) next(ctx context.Context, samples []tsdb.RefSample) (*mo
 	return &ts, entry.hash, samples[1:], nil
 }
 
-func (b *sampleBuilder) seriesGetWithRetry(ctx context.Context, sample tsdb.RefSample) (*seriesCacheEntry, bool) {
+func (b *sampleBuilder) getSeriesWithRetry(ctx context.Context, sample tsdb.RefSample) (entry *seriesCacheEntry, ok bool, err error) {
 	backoff := time.Duration(0)
-	entry, ok, err := b.series.get(ctx, sample.Ref)
 	for {
+		select {
+		case <-ctx.Done():
+			return nil, false, ctx.Err()
+		default:
+		}
+		entry, ok, err = b.series.get(ctx, sample.Ref)
 		if err == nil {
 			break
 		}
+		if _, ok := err.(unknownMetricError); ok {
```

Comment:
I'm not sure about adding such specific error taxonomy. We may end up with too many conditionals. I would ask what could lead to different handling. Is it just about recoverable and permanent? Could we call it recoverable, as in other parts of the code?